doc/computing/parallelism.rst

```diff
@@ -10,38 +10,42 @@ Parallelism, resource management, and configuration
 Parallelism
 -----------
 
-Some scikit-learn estimators and utilities can parallelize costly operations
-using multiple CPU cores, thanks to the following components:
+Some scikit-learn estimators and utilities parallelize costly operations
+using multiple CPU cores.
 
-- via the `joblib <https://joblib.readthedocs.io/en/latest/>`_ library. In
-  this case the number of threads or processes can be controlled with the
-  ``n_jobs`` parameter.
-- via OpenMP, used in C or Cython code.
+This can be either done:
 
-In addition, some of the numpy routines that are used internally by
-scikit-learn may also be parallelized if numpy is installed with specific
-numerical libraries such as MKL, OpenBLAS, or BLIS.
+- with higher-level parallelism via `joblib <https://joblib.readthedocs.io/en/latest/>`_.
+  In this case, the number of threads or processes can be controlled with the
+  ``n_jobs`` parameter.
+- with lower-level parallelism via OpenMP, used in C or Cython code.
+  In this case, parallelism is always done using threads and specifying
+  ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally
```

Comment:
That's it. It is just that `n_jobs` has currently be chosen to not be propagated to indicate the number of threads to use for OpenMP.

Personally, I think it makes sense to have `n_jobs` specify the number of threads in this case (and thus have the documentation and implementation of related user-facing API changed accordingly); but this is to be discussed among maintainers.

This document from @thomasjpfan gives some perspective on the semantic: https://thomasjpfan.github.io/parallelism-python-libraries-design/

Reply:
If setting `n_jobs` does nothing, then why does the parameter exist and why does it not raise a deprecation warning? Do you have a link to an estimator as an example? Maybe that helps to understand what is happening.

Reply:
`n_jobs` does have an effect on OpenMP parallelism. For example in `RandomSearchCV`, `joblib` will reduce the number of OpenMP threads to prevent oversubscription:

XREF: https://github.com/joblib/joblib/blob/796e7aa5c6f5a19045f06ad65415e170ae6f8e9a/joblib/parallel.py#L189

Reply:
Thanks for mentioning it.

So `n_jobs` only has an impact on the number of threads spawn by OpenMP if joblib is used? I cannot remember of a scenario where `n_jobs` specify behavior for OpenMP directly without joblib (or any of joblib's backends).

Reply:
Thanks for the details. Now it makes sense. I had understood your original comment as "the parameter exists but is never used anymore". However that isn't what happens. It is used sometimes, but there are now cases where it does nothing. As well as cases where it changes the number of threads used by OpenMP.

Sounds like overall it is quite tricky to know when `n_jobs` has an effect and if it does, what exactly it will do. Fun for the docs :D

Reply:
> So n_jobs only has an impact on the number of threads spawn by OpenMP if joblib is used? I cannot remember of a scenario where n_jobs specify behavior for OpenMP directly without joblib (or any of joblib's backends).

`n_jobs` set to control the amount of **outer** parallelism (with `loky`), for instance in `RandomSearchCV` has an indirect impact on the amount of Cython-OpenMP threading parallelism used in the inner tuned `HistGradientBoostingClassifier` or OpenBLAS threading parallelism of any estimators that relies on numpy/scipy BLAS operations for the heavy lifting inside the loky workers.

Reply:
Instead of stating:

"In this case, parallelism is always done using threads and specifying ``n_jobs`` *has no effect*"

one could add a new sentence as a new paragraph after the end of the 2 bullet point such as:

> The `n_jobs` parameters of estimators always controls the amount of parallelism managed by joblib (processes or threads depending on the joblib backend). The thread-level parallelism managed by OpenMP in scikit-learn's own Cython code or by BLAS/LAPACK libraries used by numpy and scipy operations used in scikit-learn is always controlled by environment variables or `threadpoolctl` as explained below. Note that some estimators can leverage all 3 kinds of parallelism at different points of their training and prediction methods.